feature/wilcoxon mann whitney sample size

yangwu1227 opened the pull request:
- [ ] related to #8382 
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>


**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

pep8speaks commented:
Hello @yangwu1227! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`statsmodels/stats/nonparametric.py`](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py):

> [Line 682:74](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L682): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 689:76](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L689): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 716:1](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L716): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 719:1](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L719): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 727:1](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L727): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 730:1](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L730): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 735:1](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/nonparametric.py#L735): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`statsmodels/stats/tests/test_nonparametric.py`](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py):

> [Line 502:4](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py#L502): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 504:8](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py#L504): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 510:4](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py#L510): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 511:95](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py#L511): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 512:88](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py#L512): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 551:1](https://github.com/statsmodels/statsmodels/blob/744d8c8b8632003f96f9bbcc7ac417d91097e00a/statsmodels/stats/tests/test_nonparametric.py#L551): [E304](https://duckduckgo.com/?q=pep8%20E304) blank lines found after function decorator

josef-pkt commented:
Thanks for the PR, that was fast.

a few first comments:

- Ratio is usually defined as nobs2 / nobs2
  ratio :  Sample size ratio, nobs2 = ratio * nobs1   (from a docstring)
- return should be `Holder` instance instead of dict
- line lengths look to long
- AFAIR, we always return float nobs in sample size computation, i.e. no `ceil`
- Reference in docstring missing
- I guess we should add `_onetail` to the name if we also get full power computation. e.g. https://www.statsmodels.org/dev/generated/statsmodels.stats.proportion.samplesize_proportions_2indep_onetail.html

I have not looked enough yet:

What is the alternative hypothesis here?
Can power itself be computed from or like this?

Docstring will need an example before merging. Users will not be familiar enough with this type of functions (with reference samples)

The function itself looks good and we can merge it.
However, it might be useful to split out reusable parts, e.g.
effect size from reference samples - first part of function
power or sample size computation - second part of function.

I'm thinking of whether we can include other effect size computations in the future.
For example, if we can compute it from two specified distributions, e.g. shift normal distributions, or specified ordered multinomial distributions.
It would be interesting to see the power for those cases (where also other hypothesis test could apply)

yangwu1227 commented:
Thanks for the feedback. I'll go ahead and make those changes. The game plan:

## Done

style changes:

- Renamed `ratio` to `prop_reference` to prevent confusion with the term "ratio" as it's typically defined here.
- Line lengths adjusted.

refactor stuff:

- I've updated it to return an instance of `Holder` from `statsmodels.tools.testing`. In terms of whether I should return an instance of `Holder` directly or subclass it in `statsmodels.stats.base` like `HolderTuple`, I'm flexible here—happy to go with what you recommend.
- Return floats.

## Others

logic updates 

- For adding the `add_onetail` to the name, perhaps I could introduce an `alternative=["two-sided", ...]` argument and make adjustment to `quantiles_alpha` line to handle on-sided.
  
  Interestingly, upon reviewing the reference paper, I found no specific mention of one-sided versus two-sided alternatives. The formula and their implementation only address the two-sided case. Examining the sample size formula, I see switching to a one-sided alternative would result in a smaller total sample size. My conclusion as to why they don't account for the one-sided case is to bake some conservatism into their formula, accounting for the possibility of an effect in the opposite direction.

   I could add a note in the docstring to caution users about this trade-off. Two-sided tests are more conservative and offer  buffers in the computed total sample size for effects in both directions, whereas one-sided tests result in smaller sample size required but could miss effects if the research hypothesis is wrong. This way, users can choose the option that best suits their needs.

## Regarding the Alternative Hypothesis

Based on my understanding of the paper, the alternative hypothesis is essentially left to the user’s discretion. It would be reflected in the synthetic data they generate based on their research hypothesis (e.g., a mean increase of X% relative to the reference (location shift case), a reduction in occurrence by X% compared to placebo (count distribution), a transfer of x% from the score of 1 to score of 2 (multinomial) etc.). The data passed into the calculation logic essentially encodes the alternative hypothesis, and there aren't restrictions on the underlying data.

##  Power Calculation

I guess we can back out $\beta$ and power directly from:

<img width="322" alt="Screenshot 2024-10-14 at 5 37 46 PM" src="https://github.com/user-attachments/assets/7dbe1f6c-aa06-44a0-a619-402e979416a0">

The original paper makes no mention of this application of the formula and so I could try to find if there are relevant literature on this. Formulaically, I think it's trivial to solve for power for a fixed $N$ and known constants `sigma_star`, `var_1_start`, ..., `p_star` from the provided formula.

## Future proof

Thanks for the note on extracting the logic for ranks and placements to accommodate new effect size, sample size, and power calculation methods. I might rename this function or include a note in the docstring to specify that this is one particular approach for sample size planning based on the Mann-Whitney U test formula from Edgar et al.

The two references below also seem like they could leverage some of the same rank and placement calculations:

- [Mann-Whitney U or Wilcoxon Rank-Sum Tests - Noether](https://www.ncss.com/wp-content/themes/ncss/pdf/Procedures/PASS/Mann-Whitney_U_or_Wilcoxon_Rank-Sum_Tests-Noether.pdf)
- [Tests for Two Ordered Categorical Variables - Non-Proportional Odds](https://www.ncss.com/wp-content/themes/ncss/pdf/Procedures/PASS/Tests_for_Two_Ordered_Categorical_Variables-Non-Proportional_Odds.pdf)

Mentally, I’m mapping these methods as follows:

- Noether: handles continuous data without ties.
- Non-Proportional Odds: designed for ordinal data.
- Edgar et al.: more general, able to handle arbitrary underlying data.

josef-pkt commented:
we have RankCompareResult.test_prob_superior for the one-sided alternative
https://www.statsmodels.org/dev/generated/statsmodels.stats.nonparametric.RankCompareResult.test_prob_superior.html

example would be if we want to test whether a new treatment is "better" than an old treatment, and not just testing whether they are different (under the alternative)

So. it would be useful to have the explicit option for alternatives, two-sided, smaller, larger.

Aside: I don't remember this
test_prob_superior allows for a "non-zero margin", i.e. a null hypothesis different from equality p=0.5.
But no I don't remember whether the variances are correct for this, or the variances are derived under p=0.5.
In the second case, the docstring should mention this, and it would most likely only be a good approximation for p not too far from 0.5.

josef-pkt commented:
"superiority" is confusing here because it has two different meanings

- p > 0.5 means distribution 1 is stochastically superior (in the definition of the rank compare hypothesis), i.e. has "larger values" 
- for general cases superiority test means one-sided alternative, possibly with a non-zero margin, e.g. mean of distribution of sample 1 is at least d units larger than the mean of sample 2,   mu1 > mu2 + d,  for example if we want to show that a new treatment is really better. Or mu1 > mu2 - d, if we want to show that treatment one is at least almost as good as (at most d worse than) the reference.
  non-inferiority test are similar. (definitions are not all the same across references)

`RankCompareResult.test_prob_superior` refers to the second meaning of "superior", as used in other hypothesis tests

AFAICS, the links in your last comment refer to the first definition, which is the statistic in MWM and brunner-munzel. 
The literature is a bit confusing because the same statistic has been reinvented several times under different names.
I tried to describe it a bit in the Notes section of
https://www.statsmodels.org/dev/generated/statsmodels.stats.nonparametric.rank_compare_2indep.html
(I called the function `rank_compare` to avoid a bit the confusing names)

yangwu1227 commented:
I’ve refactored the power calculation function by breaking out the ranks and placement calculations into a helper function. Hopefully this could be useful. I also added support for the alternative parameter to handle both one-tailed and two-tailed tests.

I threw in some extra notes in the docstring too, covering ranks, placements, the reference vs. synthetic, and one-sided vs. two-sided tests.

Any feedback would be greatly appreciated.

And big thanks for all the knowledge sharing and resources you’ve shared. Hoping I’ll have more time to work and dig into this kind of stuff next fall... if my Ph.D. application is successful, fingers crossed haha.

josef-pkt commented on statsmodels/stats/tests/test_nonparametric.py:
```diff
@@ -493,3 +495,277 @@ def test_rank_compare_vectorized():
+
+@pytest.mark.parametrize(
+    "x1, x2, expected_holder",
+    [
```
To improve readability, I usually define the list of cases in a separate list.

The advantage is also that sometimes I use then the list in several parameterized unit test, e.g. if those are added later

yangwu1227 replied:
Just to make sure I handle this correctly. I'm thinking of doing the following: 

```python
@pytest.fixture(scope="module")
def two_sample_continuous():
    """
    This fixture generates two samples of
    continuous data without ties.
    """
    x1 = np.array([1.1, 2.2, 3.3, 4.4, 5.5])
    x2 = np.array([6.6, 7.7, 8.8, 9.9, 10.1])
    return x1, x2

@pytest.fixture(scope="module")
def two_sample_ordinal():
    """
    This fixture generates two samples of
    ordinal data with ties.
    """
    x1 = np.array([1, 1, 2, 2, 3])
    x2 = np.array([4, 5, 6, 7, 8])
    return x1, x2
   
# These can be used in parametrized tests like

```python
@pytest.mark.parametrize(
    "data_samples, expected_holder",
    [
        # No ties continuous data
        (
            "two_sample_continuous",
             ...,
        ),
        # Ties ordinal data
        (
            "two_sample_ordinal",
            ...,
        ),
    ],
)
def test_compute_rank_placements(request, data_samples, expected_holder):
    # Get the fixture values
    x1, x2 = request.getfixturevalue(data_samples)
```

Do you think this does the trick? Or perhaps a simpler set up is to create these as a lists of two tuples(x1, x2) at the top of the module directly without making them fixtures.

josef-pkt replied:
It's simpler just to use a list
(I never used fixtures. I never found them necessary for this.)

for example
https://github.com/statsmodels/statsmodels/blob/main/statsmodels/stats/tests/test_rates_poisson.py#L108
and many other cases in that module

and two lists of cases can be concatenated with just `+` in parametrize call

josef-pkt commented on statsmodels/stats/nonparametric.py:
```diff
@@ -662,3 +663,273 @@ def cohensd2problarger(d):
+        The type I error rate for the test (two-sided).
+    power : float
+        The desired power of the test.
+    prop_reference : float, optional
```
still confusing and not consistent with other power functions

use "nobs_ratio"  in signature
and compute and use prop_reference internally so that computations don't need to change

yangwu1227 replied:
This is done.

josef-pkt commented on statsmodels/stats/nonparametric.py:
```diff
@@ -662,3 +663,273 @@ def cohensd2problarger(d):
+    ----------
+    reference_sample : array_like
+        Advance information for the reference group.
+    synthetic_sample : array_like
```
better reverse the sequence here between reference and sythetic sample.
`rank_compare_sample_size(synthetic_sample, reference_sample, ...)`
    

In my convention for 2 sample test, the reference case is usual the last
x1, x2    with x1 compared to reference x2, e.g. effect = mean1 - mean2
or
`statsmodels.stats.nonparametric.rank_compare_2indep(x1, x2, ...)`

yangwu1227 replied:
I reversed everything, including tests, docstrings, and examples, and renamed `nobs1` and `nobs2` to `nobs_ref` and `nobs_syn` for greater clarity. Even I got a bit confused when updating trying to remember which is 1 and which is 2. So users may be too. 

This time it will be  easier to read and ensures that users won’t have to remember which parameter corresponds to which group; it will be reflected in the attribute names.

bashtage commented:
Sorry for a small hijack. @josef-pkt There have been a couple of emails relating to statsmodels, one from me and one from @stefanv . If you haven't received them, could you send me an email from a good address?

yangwu1227 commented:
I’ve taken care of the refactoring as suggested. Happy to do more if there are conventions or practices that better align with project's standards. Thanks for the feedback. 

Regarding the function name, I may not have enough context to offer the best suggestion here, so I’ll defer to your expertise and knowledge of the project’s future plans. I much agree that the reference/synthetic sample setup in this formula is quite specific to the paper. Although Noether's formula (referenced in the PASS article) also discusses using simulated data for effect size estimation, it ultimately relies on an estimated effect size input directly rather than this setup. Given that, perhaps a name like `samplesize_wmw_[something specific to this method]` could provide flexibility for adding different formulas in the future.

On a related note, if aditional methods are added in the future that differ in varios ways, I’m reminded (in my own case at least) of the helpfulness of a visual guide like [scikit-learn's machine learning map](https://scikit-learn.org/1.3/tutorial/machine_learning_map/). A similar map for sample size and power calculation tools could be valuable, organized by factors such as data type, parametric vs. non-parametric methods, handling of ties, etc. By no means a trivial undertaking haha, though your excellent point on the importance of naming got me thinking perhaps such a resource could be useful if more tools are added.

josef-pkt commented:
overview of hypothesis tests, sample size and power are not available for all test:

I have seen overviews in table form but not as graphical decision tree or flow chart.
brief google search:
this is the largest that I found https://statsandr.com/blog/files/overview-statistical-tests-statsandr.pdf
and https://www.pnrjournal.com/index.php/home/article/view/18/16
There are others with a different emphasis like https://www.brookes.ac.uk/getmedia/bede726d-771d-461f-900b-a3526fc7e199/Stats-Flow-Chart.pdf

a Google photos search for "flowchart for commonly used statistical tests" finds many more.

a full list is pretty large, e.g. we also have hypothesis tests and related functions for count data in `rates` which is often not included as separate data type. Which tests are important also depends on the scientific field.
#8331 overview issue that I started but mainly to see what we have or scipy has and what is still missing.
(I'm usually more focused on completeness or details and not on what would be a good learning map.) 

And for each test, we would like to have as far as possible:

- standard hypothesis test
- confidence interval
- equivalence test
- superiority/noninferiority tests
- power
- sample size
- and possibly others like prediction interval, tolerance interval, standard effect size if available

(aside: prediction interval, tolerance interval are useful for outlier testing and quality control charts. Standard effect sizes are popular in psychology and similar fields.)

josef-pkt commented:
unrelated test failure
100% unit test coverage according to codecov

ok, to finish this up
rename function to `samplesize_rank_compare_onetail` and I will merge

The function looks good as it is.

There are some open design question which are not clear to me right now.
Not clear yet whether `onetail` sample size function should have alternative "larger" and "smaller" or "onesided" or both.
(currently I'm think of both, i.e. have 3 one-sided alternative options)
I have not looked at possible code reuse. for example https://www.statsmodels.org/dev/generated/statsmodels.stats.power.normal_sample_size_one_tail.html

yangwu1227 commented:
Thanks for sharing the design philosophy overview. I like the 3 one-sided alternative options. Perhaps something like:

```python
if alternative == "two-sided":
    quantile_prob = 1 - alpha / 2
elif alternative in ["one-sided", "larger"]:
    quantile_prob = 1 - alpha  
elif alternative == "smaller":
    quantile_prob = alpha  
```

And just explain in the docsring that "one-sided" defaults to "larger".

josef-pkt commented:
one problem for the larger, smaller alternative is that sample size only exists if the alternative is in the correct tail.
if the user specifies reference samples with p < 0.5, then we cannot find a sample size for an alternative that p > 0.5  (or something like that.
If the alternative is "one-sided", then we assume the user meant that the alternative (larger or smaller) is in the direction of where the sample size exists and can be calculated. The "one-sided" is easier to use, but users might make mistakes and get the wrong larger or smaller alternative.

> This means that we don't need to distinguish between the two directions when calculating quantile_prob

This should not be the case if we really have "larger" and "smaller" alternatives. They need to fail if the sample size does not exist when we are in the wrong tail (always power < alpha)

Your implementation looks correct for "one-sided" alternative and automatically chooses the appropriate tail to get a sample size.
But I did not look in details about where the incorrect "sign" would show up. 

recent comment: https://github.com/statsmodels/statsmodels/issues/9378#issuecomment-2389479221
So this is not yet cleanly implemented for the samplesize_xxx functions

yangwu1227 commented:
Appreciate the detailed explanation of the misaligned effect sizes and alternative hypotheses issue. I see a primary concern is ensuring that the direction of the `relative_effect` aligns with the specified alternative ("larger" or "smaller") if we allow usee to explicitly specify those options. If there's misalignment, then sample size calculation is invalid (we are in the wrong tail), and the power cannot be achieved under these conditions.

Based on your comments

> We should find explicit solution to intercept impossible cases at least for computing sample size.

This seems like a good prudent approach to me. To be make the implementation cleaner, I could update the `alternative` parameter to accept "two-sided", "larger", and "smaller". Then, after calculating `relative_effect`, check that:

* for "larger", `relative_effect > 0.5`
* for "smaller", `relative_effect < 0.5`
* I’m now wondering if keeping a generic `one-sided` option poses more risks than benfits

And a misalignment is detected, raise a clear exception informing the user of the issue.

josef-pkt commented:
> I’m now wondering if keeping a generic one-sided option poses more risks than benefits

It's not clean, but easier to understand.
Thinking about larger and smaller always requires effort in which direction the alternative and rejection region should go.

(In the early days of statsmodels we had hypothesis tests with "one-sided" alternative, but I stopped that because it was not clear and clean enough, leaving it open to misuse. However, "larger" and "smaller" can also lead to confusion and misunderstandings.) 

For power computation "one-sided" is not an option, because we can compute power even if it's smaller than alpha. Being in the "wrong" tail is not a problem for the definition nor computation.  
`sample_size_xxx_onetail` is the only type of functions where "one-sided" can be justified.

If you make the changes in your last comment, than this would be clean and we can post-pone the decision to add "one-sided" option.
In the same way, you could raise an exception when relative effect is 0.5, instead of adding the fudge term in the computation (denominator).

yangwu1227 commented:
For the exception, do you prefer that I create a  e.g., `SampleSizeCalculationError` custom exception in `statsmodels.tools.sm_exceptions` or just a regular `ValueError`? thanks.

josef-pkt commented on statsmodels/stats/nonparametric.py:
```diff
@@ -662,3 +663,273 @@ def cohensd2problarger(d):
+    """
+    Compute the required sample size for the non-parametric
+    Mann-Whitney U test.
+
```
needs empty line before second sentence "This function implements ..."

first line should be a summary sentence that fits on one line.

josef-pkt commented:
21 commits in this PR are too many. Needs squashing before merge.

We might want to preserve the changes with the discussion here.
You could open a new PR with a rebased, squashed version of this and I will merge that.

Thanks for the PR and working on this.

yangwu1227 commented:
Just opened a new PR with squashed commits. 

Thanks for all feedback and guidance.

josef-pkt commented:
why is this marked as merged?

new github feature?

yangwu1227 commented:
> why is this marked as merged? new github feature?
> 
> No, both PRs are the same. This PR is also squashed.

That’s my fault, and I apologize. I misunderstood your original request and ended up syncing both feature branches. In hindsight, I should have  -> new local branch -> merge the changes there -> and then squash them.
